### ISSUE TITLE
Add support for Digitone II

### ISSUE THREAD

**Issue by szszoke** [dagargo/elektroid#165]
Below is the device definition

```json
{
  "id": 43,
  "name": "Elektron Digitone II",
  "alias": "dn2",
  "filesystems": 56,
  "storage": 0
}
```

Listing directories seem to work. I tried downloading a preset and that failed.

Parsing at `package.c:536` fails because of an invalid JSON string: `{\"metaversion\": 1,}`. There is a trailing comma that should not be there. There could be other places where the returned JSON is invalid.

I am also getting a `GLib-WARNING`:

```bash
GLib-WARNING **: 12:32:09.515: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: <data>:1:18: Parse error: unexpected character `,', expected character `]'
```

**Comment by szszoke** [comment]
I added a quick and dirty workaround which seems to solve the issue. I was able to download the factory preset and upload it to a new slot.

```diff
@@ -532,6 +532,13 @@ package_receive_pkg_resources (struct package *pkg,
 
   control->part++;
 
+  if (metadata_file.content->len > 2
+    && metadata_file.content->data[metadata_file.content->len - 1] == '}'
+    && metadata_file.content->data[metadata_file.content->len - 2] == ',')
+    {
+    g_byte_array_remove_index(metadata_file.content, metadata_file.content->len - 2);
+  }
+
   parser = json_parser_new ();
   if (!json_parser_load_from_data (parser,
                                   (gchar *) metadata_file.content->data,
```

**Comment by dagargo** [comment]
Just a side note.

All the filesystems in Elektroid are named after the objects that the different devices offer.
Sometimes, a device calls the programmed sounds presets but sounds or programs are used too. This depends on the vendors and the products but Elektroid always uses the name as it appears there.

By the way, the name has moved from `Sound` to ` Preset` in the Digitone II as it happened with Digitakt II. Probably we should rename the filesystem for the Digitone II as we did with the Digitakt II but other than the name there are no differences.

**Comment by dagargo** [comment]
As a reminder, a Digitone I offers 3 filesystems:

* FS_DATA_ANY (native filesystem only usable from the CLI and used for testing)
* FS_DATA_PRJ
* FS_DATA_SND

But you've said these 2 things and I got lost.

> I tried downloading a preset and that failed.

> Sounds do not seem to be affected. 

So, what is failing? Projects or presets (sounds)?

**Comment by dagargo** [comment]
Probably, the storage for the +Drive could be used too. I'm leaving here this for later.

**Comment by szszoke** [comment]
Sorry. You're right. Let me rephrase that. I tried to download the factory project called `PRESET`. It failed because of the invalid JSON.

I then switched the drop-down to `Sounds` but downloading those worked without my quick fix.

I need to look into the Elektron Transfer app. I mostly back up full projects so I'm that familiar with the other file systems.

**Comment by szszoke** [comment]
> Probably, the storage for the +Drive could be used too. I'm leaving here this for later.

I used the existing Digitone entry as my template. I'm not quite sure what the +Drive bit would mean so I left it out.

**Comment by dagargo** [comment]
I've just pushed 1f0d6cf29966d4a0913ac553c2470512b0be4e19 to fix the aforementioned GLib warning. A variable was not properly initialized.

**Comment by dagargo** [comment]
Perhaps, setting `storage` to `1` could show something.

As all devices have a +Drive. or so I think, the first row should be present.

```
$ elektroid-cli df 1
Storage                         Size            Used       Available       Use%
+Drive                      959.5MiB        281.9MiB        677.6MiB     29.38%
RAM                            64MiB        789.2KiB        63.23MiB      1.20%
```

These values are shown at the bottom right of the GUI too.

**Comment by szszoke** [comment]
> I've just pushed [1f0d6cf](https://github.com/dagargo/elektroid/commit/1f0d6cf29966d4a0913ac553c2470512b0be4e19) to fix the aforementioned GLib warning. A variable was not properly initialized.

That fixed the segfault that was happening before. Now I just get this error and the download then finishes:

```bash
ERROR:connectors/package.c:540:package_receive_pkg_resources: Unable to parse stream: <data>:1:18: Parse error: unexpected character `,', expected character `]'. Continuing...
```

> Regarding the error when downloading the PRESET project, does this happen with other projects or only with this one?

It happens with the factory `PRESETS` project and my own projects as well.

> Perhaps, setting storage to 1 could show something.

It doesn't seem to work. Looks like the Digitone is not responding with anything. By the way, the CLI locks up to the point where `CTRL+C`, `killall` or `kill <PID>` just doesn't work to close it. I had to use `kill -9 <PID>`.

```bash
~ ./src/elektroid-cli df 1 -vv
DEBUG:preferences.c:233:preferences_load: Loading preferences from '/home/<REDACTED>/.config/elektroid/preferences.json'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'localDir'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'remoteDir'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'showRemote'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'autoplay'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'mix'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'showGrid'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'gridLength'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'playSampleWhileLoading'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'audioBufferLength'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'stopDeviceWhenConnecting'...
DEBUG:preferences.c:242:preferences_load: Loading preference 'elektronLoadSoundTags'...
DEBUG:backend_alsa.c:369:backend_get_system_subdevices: Adding hw:3 (name 'Elektron Digitone II', subname 'Elektron Digitone II MIDI 1')...
DEBUG:backend.c:313:backend_init_midi: Initializing backend (ALSA) to 'hw:3,0,0'...
DEBUG:backend_alsa.c:87:backend_init_int: Setting blocking mode...
DEBUG:backend.c:634:backend_rx_drain: Draining buffers...
DEBUG:backend.c:403:backend_rx_raw_loop: Timeout (1000)
DEBUG:backend.c:326:backend_init_midi: Stopping device...
DEBUG:backend.c:706:backend_init_connector: Connector elektron matches the device
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (6): f0 7e 7f 06 01 f7
DEBUG:backend.c:614:backend_rx_sysex: Raw message received (6): f0 7e 00 7c 00 f7
DEBUG:backend.c:148:backend_midi_handshake: Illegal SUB-ID2
DEBUG:backend.c:746:backend_init_connector: Testing elektron connector (non standard handshake)...
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (13): f0 00 20 3c 10 00 00 00 00 00 00 01 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (5): 00 00 00 00 01
DEBUG:backend.c:614:backend_rx_sysex: Raw message received (54): f0 00 20 3c 10 00 04 00 64 00 00 01 2b 16 00 01 02 03 04 06 07 09 00 50 52 51 53 54 55 56 00 57 58 59 5a 5b 5c 5d 00 5e 44 69 67 69 74 6f 00 6e 65 20 49 49 00 f7
DEBUG:connectors/elektron.c:818:elektron_rx: Message received (41): 00 64 00 00 81 2b 16 01 02 03 04 06 07 09 50 52 51 53 54 55 56 57 58 59 5a 5b 5c 5d 5e 44 69 67 69 74 6f 6e 65 20 49 49 00
DEBUG:connectors/elektron.c:3313:elektron_configure_device: Using /home/<REDACTED>/.config/elektroid/elektron/devices.json...
DEBUG:connectors/elektron.c:3362:elektron_configure_device: Device 43 found
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (13): f0 00 20 3c 10 00 00 00 01 00 00 02 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (5): 00 01 00 00 02
DEBUG:backend.c:614:backend_rx_sysex: Raw message received (26): f0 00 20 3c 10 00 04 00 65 00 01 02 30 30 00 32 38 00 31 2e 30 30 00 41 00 f7
DEBUG:connectors/elektron.c:818:elektron_rx: Message received (16): 00 65 00 01 82 30 30 32 38 00 31 2e 30 30 41 00
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (13): f0 00 20 3c 10 00 00 00 02 00 00 03 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (5): 00 02 00 00 03
DEBUG:backend.c:614:backend_rx_sysex: Raw message received (18): f0 00 20 3c 10 00 05 00 66 00 02 03 26 55 60 44 69 f7
DEBUG:connectors/elektron.c:818:elektron_rx: Message received (9): 00 66 00 02 83 26 d5 c4 e9
DEBUG:connectors/elektron.c:3509:elektron_handshake: UID: e9c4d526
DEBUG:backend.c:776:backend_init_connector: Using elektron connector...
Storage                         Size            Used       Available       Use%
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (14): f0 00 20 3c 10 00 00 00 03 00 00 05 01 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (6): 00 03 00 00 05 01
DEBUG:backend.c:403:backend_rx_raw_loop: Timeout (5000)
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (14): f0 00 20 3c 10 00 00 00 04 00 00 05 02 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (6): 00 04 00 00 05 02
DEBUG:backend.c:403:backend_rx_raw_loop: Timeout (5000)
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (14): f0 00 20 3c 10 00 00 00 05 00 00 05 03 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (6): 00 05 00 00 05 03
DEBUG:backend.c:403:backend_rx_raw_loop: Timeout (5000)
DEBUG:backend_alsa.c:206:backend_tx_sysex_internal: Raw message sent (14): f0 00 20 3c 10 00 00 00 06 00 00 05 04 f7
DEBUG:connectors/elektron.c:773:elektron_tx: Message sent (6): 00 06 00 00 05 04
```

**Comment by szszoke** [comment]
I checked with the Elektron Transfer app and it doesn't show any info about the storage so this is probably not implemented by the device.

**Comment by dagargo** [comment]
Let's forget about the filesystems then.

> I did send Elektron an email about this and they said they will fix it so maybe your fix is enough?

I'd say so. If it happens with any project, they are gonna fix it.

**Comment by szszoke** [comment]
Well, it doesn't seem to affect Elektron Transfer but they might be doing the same thing as you.

**Comment by dagargo** [comment]
Probably, the `.metadata` file is not needed in the Digitone case as the only interesting thing there are the samples.

The code doesn't do anything for metadata files that have no `sample_references` member. That's why it works. If Elektron doesn't solve it, it's not really an issue.

```
$ elektroid-cli elektron-data-dl 2:/projects/2/.metadata
$ jq . 'PROJECT.dtdata.metadata'
{
  "metaversion": 1,
  "sample_references": [
    {
      "hash": 1751731247,
      "size": 900
    },
    {
      "hash": 2954592109,
      "size": 928
    },
    {
      "hash": 1940590663,
      "size": 98540
    },
    {
      "hash": 1949080751,
      "size": 1708
    },
    {
      "hash": 3939968925,
      "size": 46552
    },
    {
      "hash": 2762110027,
      "size": 59008
    },
    {
      "hash": 2571488959,
      "size": 15248
    },
    {
      "hash": 3955730059,
      "size": 21570
    },
    {
      "hash": 854823783,
      "size": 73662
    },
    {
      "hash": 3970949489,
      "size": 19684
    },
    {
      "hash": 118647361,
      "size": 37722
    }
  ]
}
```

**Comment by szszoke** [comment]
Maybe Transfer is not even attempting to download it?

**Comment by dagargo** [comment]
That's very likely.

**Comment by szszoke** [comment]
Would it make sense to infer from the filesystems if metadata should be downloaded?

**Comment by dagargo** [comment]
Definitely, it can be done. The problem is that that function is used by all the filesystems so it must be changed with special care.

I need to think about it.

**Comment by dagargo** [comment]
I've pushed a new branch `feat-add-digitone-ii` to work on this.

Commit 2a824131d2e821d17386be78c6ffbf2818c7366b avoids metadata downloading if not needed so this will fix the download error.

Commit ff387151bbe9f25691ec45c02530639e002bb185 replaces your proposed sound filesystem with the preset filesystem. Regarding their functionalities and implementation, they are both almost identical, only the extensions differ.

With regard to the extensions, check this two functions in case the new Digitone II is backwards compatible (Digitakt II is) with the Digitone (I) files as the old extensions will need to be added.

https://github.com/dagargo/elektroid/blob/ff387151bbe9f25691ec45c02530639e002bb185/src/connectors/elektron.c#L2961-L2962

https://github.com/dagargo/elektroid/blob/ff387151bbe9f25691ec45c02530639e002bb185/src/connectors/elektron.c#L2974-L2975

**Comment by dagargo** [comment]
I've just read that they seem to be compatible too!

https://www.elektronauts.com/t/dn1-sysex-into-dnii-compatibility-issues/222478

**Comment by szszoke** [comment]
I will give your changes a try over the weekend.

Regarding the extensions, would it make sense to move them to the device definitions in a more explicit way than the device alias?

I.e an "extensions" object where each key represents a file type and the values are arrays of strings.

We could then say that the first extension for a given file type is the default one and would be used for saving files on the computer.

If a device supports multiple extensions then we could just add it to the list and Elektroid would allow uploading files with those extensions.

For the older devices we would only have a single extension for each file type.

For the second generation devices we would have both the new and the old file extensions.

**Comment by szszoke** [comment]
Something broke.

Downloading a project results in a segmentation fault. It happens on `master` too.

```
(elektroid:20034): GLib-CRITICAL **: 15:44:33.278: g_atomic_ref_count_dec: assertion 'old_value > 0' failed
```

```bt
#0  0x00007fffb7e00010 in ?? ()
#1  0x00007ffff7235a9c in array_free (array=0x7fffe95ffb30, flags=(FREE_SEGMENT | PRESERVE_WRAPPER)) at ../glib/glib/garray.c:541
#2  0x00007ffff7238ab3 in g_array_free (farray=<optimized out>, free_segment=free_segment@entry=1) at ../glib/glib/garray.c:525
#3  0x00007ffff7238b4a in g_byte_array_free (array=<optimized out>, free_segment=free_segment@entry=1) at ../glib/glib/garray.c:2792
#4  0x0000555555568626 in idata_free (idata=idata@entry=0x7fffe95ffa50) at utils.c:525
#5  0x00005555555731de in package_receive_pkg_resources (pkg=pkg@entry=0x7fffe95ffb30, payload_path=0x7fffc000fe00 "/4", control=control@entry=0x5555555a8640 <tasks>, backend=backend@entry=0x5555555a8a40 <backend>, download_data=download_data@entry=0x555555571000 <elektron_download_data_prj>, 
    download_sample=download_sample@entry=0x55555556d820 <elektron_download_sample_part>, type=PKG_FILE_TYPE_DATA_PROJECT) at connectors/package.c:679
#6  0x000055555556ddfc in elektron_download_pkg (backend=0x5555555a8a40 <backend>, path=<optimized out>, output=0x7fffe95ffbe0, control=0x5555555a8640 <tasks>, type=type@entry=PKG_FILE_TYPE_DATA_PROJECT, ops=ops@entry=0x5555555a5200 <FS_DATA_PRJ_OPERATIONS>, download=0x555555571000 <elektron_download_data_prj>)
    at connectors/elektron.c:2656
#7  0x000055555556deee in elektron_download_data_prj_pkg (backend=<optimized out>, path=<optimized out>, pkg=<optimized out>, control=<optimized out>) at connectors/elektron.c:3549
#8  0x000055555558d95c in elektroid_download_task_runner (userdata=<optimized out>) at elektroid.c:1704
#9  0x00007ffff72a6026 in g_thread_proxy (data=0x555555d30330) at ../glib/glib/gthread.c:892
#10 0x00007ffff6a2939d in start_thread (arg=<optimized out>) at pthread_create.c:447
#11 0x00007ffff6aae49c in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

**Comment by szszoke** [comment]
I guess this actually makes sense because you never actually do anything with `metadata_file` for non-sample based devices.

**Comment by szszoke** [comment]
Manually initializing the members of `metadata_file` resolves the segfault:

```diff
diff --git a/src/connectors/package.c b/src/connectors/package.c
index 084d140..d203aa7 100644
--- a/src/connectors/package.c
+++ b/src/connectors/package.c
@@ -513,7 +513,11 @@ package_receive_pkg_resources (struct package *pkg,
   gchar *sample_path, *metadata_path;
   struct package_resource *pkg_resource;
   GString *package_resource_path;
-  struct idata metadata_file, payload_file, sample_file, file;
+  struct idata metadata_file = {
+    .content = NULL,
+    .name = NULL,
+    .info = NULL
+  }, payload_file, sample_file, file;
   struct elektron_data *data = backend->data;
 
   //Analog Rytm, Digitakt, Analog Rytm MKII, Model:Samples and Digitakt II
```

**Comment by szszoke** [comment]
I am not a C programmer so I don't know what is idiomatic in this case but I would think it is good practice to always use some known initial value.

The code below means to do that.

```diff
diff --git a/src/connectors/package.c b/src/connectors/package.c
index 084d140..61e97b3 100644
--- a/src/connectors/package.c
+++ b/src/connectors/package.c
@@ -516,6 +516,11 @@ package_receive_pkg_resources (struct package *pkg,
   struct idata metadata_file, payload_file, sample_file, file;
   struct elektron_data *data = backend->data;
 
+  idata_init(&metadata_file, NULL, NULL, NULL);
+  idata_init(&payload_file, NULL, NULL, NULL);
+  idata_init(&sample_file, NULL, NULL, NULL);
+  idata_init(&file, NULL, NULL, NULL);
+
   //Analog Rytm, Digitakt, Analog Rytm MKII, Model:Samples and Digitakt II
   if (type == PKG_FILE_TYPE_DATA_PROJECT && data->device_desc.id != 8 &&
       data->device_desc.id != 12 && data->device_desc.id != 16 &&
```

**Comment by szszoke** [comment]
Sending a DN1 `Project`/`Sound` to a DN2 works. The project appears on the DN2 with a special marking and when you open it it says `Importing`. The Sound just appears as a regular `Preset`.

Downloading of the same items from the DN2 fails however.

```diff
diff --git a/src/connectors/elektron.c b/src/connectors/elektron.c
index 216b065..10a0609 100644
--- a/src/connectors/elektron.c
+++ b/src/connectors/elektron.c
@@ -2964,10 +2964,24 @@ elektron_get_dev_exts_prj (struct backend *backend,
 {
   struct elektron_data *data = backend->data;
   GSList *exts = elektron_get_dev_exts (backend, ops);
-  if (data->device_desc.id == 42)	//Digitakt II
+
+  switch (data->device_desc.id)
     {
-      exts = g_slist_append (exts, strdup ("dtprj"));
+      //Digitakt II
+    case 42:
+      {
+	exts = g_slist_append (exts, strdup ("dtprj"));
+	break;
+      }
+
+      //Digitone II
+    case 43:
+      {
+	exts = g_slist_append (exts, strdup ("dnprj"));
+	break;
+      }
     }
+
   return exts;
 }
 
@@ -2977,10 +2991,24 @@ elektron_get_takt_ii_pst_exts (struct backend *backend,
 {
   struct elektron_data *data = backend->data;
   GSList *exts = elektron_get_dev_exts (backend, ops);
-  if (data->device_desc.id == 42)	//Digitakt II
+
+  switch (data->device_desc.id)
     {
-      exts = g_slist_append (exts, strdup ("dtsnd"));
+      //Digitakt II
+    case 42:
+      {
+	exts = g_slist_append (exts, strdup ("dtsnd"));
+	break;
+      }
+
+      //Digitone II
+    case 43:
+      {
+	exts = g_slist_append (exts, strdup ("dnsnd"));
+	break;
+      }
     }
+
   return exts;
 }
```

I think that especially now that there are two devices that are confirmed to be backwards compatible with old file formats, the identifiers of those devices should not be hard-coded into the source code.

Example

```json
{
  "id": 43,
  "name": "Elektron Digitone II",
  "alias": "dn2",
  "filesystems": 280,
  "storage": 0,
  "extensions": {
    "project": [
      "dn2prj",
      "dnprj"
    ],
    "preset": [
      "dn2pst",
      "dnsnd"
    ]
  }
}
```

**Comment by szszoke** [comment]
I will investigate this further later.

**Comment by dagargo** [comment]
> Something broke.
Downloading a project results in a segmentation fault.

If this really happens in master, we do need to take a look at this before anything else. I'll take a look at your findings and do some tests during the following days.

**Comment by szszoke** [comment]
> > Something broke.
> Downloading a project results in a segmentation fault.
> 
> If this really happens in master, we do need to take a look at this before anything else. I'll take a look at your findings and do some tests during the following days.

That might have been my error. But I will check again later.

**Comment by szszoke** [comment]
Actually it does happen on `master` but for a slightly different reason. It's because of this:

```bash
ERROR:connectors/elektron.c:2291:elektron_open_datum: Operation not permitted (Error: Could not resolve path)
```

`elektron_download_data_prefix` returns early because of it and then there is a `goto get_payload` directly after that gets executed due to the error.

The offending path is `/<DEVICE_NO>/.metadata`.

**Comment by szszoke** [comment]
Downloading the DN1 project produces this log:

```bash
ERROR:connectors/elektron.c:2291:elektron_open_datum: Operation not permitted (Read: Permission denied)
ERROR:connectors/package.c:689:package_receive_pkg_resources: Error while downloading payload
ERROR:elektroid.c:1713:elektroid_download_task_runner: Error while downloading
```

`Read: Permission denied` seems to come from the device itself.

**Comment by szszoke** [comment]
Apparently this is expected behavior: https://www.elektronauts.com/t/digitone-ii-bug-reports/222351/88

**Comment by dagargo** [comment]
Thanks for taking the time to investigate this issue. :+1: 

So, is everything working then?

**Comment by szszoke** [comment]
No worries! Everything is working as much as it is working with Elektron Transfer, if the code that initializes things to `NULL` is present.

**Comment by dagargo** [comment]
Merged into master in #168.
